services/fileuploader/fileuploader.go

```diff
@@ -97,9 +112,21 @@ func (p *provider) updateLoop(ctx context.Context, backendConfig backendconfig.B
 	ch := backendConfig.Subscribe(ctx, backendconfig.TopicBackendConfig)
 
 	for ev := range ch {
-		settings := make(map[string]StorageSettings)
+		p.mu.RLock()
+		currentSettingsMap := p.storageSettings
+		currentFileManagerMap := p.fileManagerMap
+		p.mu.RUnlock()
+		settingsMap := make(map[string]StorageSettings)
+		filemanagerMap := make(map[string]func() (filemanager.FileManager, error))
 		configs := ev.Data.(map[string]backendconfig.ConfigT)
 		for workspaceId, c := range configs {
+			currentWorkspaceConfig, ok := currentSettingsMap[workspaceId]
+			// no change in workspace config, don't process the same config again
+			if ok && !c.UpdatedAt.After(currentWorkspaceConfig.updatedAt) {
+				settingsMap[workspaceId] = currentWorkspaceConfig
+				filemanagerMap[workspaceId] = currentFileManagerMap[workspaceId]
```

Comment:
Won't we have a race anyway here? You're getting a map which works via address so if the content of the map is changed in line 94, you would still have a read/write race condition here. A test with `-race` should highlight this.